server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java

```diff
@@ -89,56 +83,59 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
-                for (final ObjectObjectCursor<String, List<AliasMetaData>> cursor : aliasMap) {
+                final Set<String> returnedAliasNames = new HashSet<>();
+                for (final ObjectObjectCursor<String, List<AliasMetaData>> cursor : response.getAliases()) {
                     for (final AliasMetaData aliasMetaData : cursor.value) {
-                        aliasNames.add(aliasMetaData.alias());
                         if (namesProvided) {
+                            // display only indices with no aliases
                             indicesToDisplay.add(cursor.key);
                         }
+                        returnedAliasNames.add(aliasMetaData.alias());
                     }
                 }
-
-                // first remove requested aliases that are exact matches
-                final SortedSet<String> difference = Sets.sortedDifference(Arrays.stream(aliases).collect(Collectors.toSet()), aliasNames);
-
-                // now remove requested aliases that contain wildcards that are simple matches
-                final List<String> matches = new ArrayList<>();
-                outer:
-                for (final String pattern : difference) {
-                    if (pattern.contains("*")) {
-                        for (final String aliasName : aliasNames) {
-                            if (Regex.simpleMatch(pattern, aliasName)) {
-                                matches.add(pattern);
-                                continue outer;
+                // compute explicitly requested aliases that have not been found
+                final SortedSet<String> missingAliases = new TreeSet<>();
+                for (int i = 0; i < aliases.length; i++) {
+                    if (MetaData.ALL.equals(aliases[i]) || Regex.isSimpleMatchPattern(aliases[i]) || aliases[i].charAt(0) == '-') {
+                        // only explicitly requested aliases will be returning 404
+                        continue;
+                    }
+                    int j = i + 1;
+                    for (; j < aliases.length; j++) {
+                        if (aliases[j].charAt(0) == '-'
+                                && (Regex.isSimpleMatchPattern(aliases[j].substring(1)) || MetaData.ALL.equals(aliases[j].substring(1)))) {
+                            // this is an exclude pattern
+                            if (Regex.simpleMatch(aliases[j].substring(1), aliases[i]) || MetaData.ALL.equals(aliases[j].substring(1))) {
+                                break;
                             }
                         }
                     }
+                    if (j == aliases.length) {
+                        // explicitly requested alias not excluded by any "-" wildcard in expression
+                        if (false == returnedAliasNames.contains(aliases[i])) {
+                            missingAliases.add(aliases[i]);
+                        }
+                    }
```

Comment:
I am not following what changes were made here, can you maybe explain? I was expecting changes needed only in IndicesAndAliasesResolver to add support for pattern exclusion, why do we need to make changes to the transport action too?

Reply:
what you explain sounds like a bug in the current wildcard exclusion implementation, besides the fact that security does not support pattern exclusions. Is that correct?

Reply:
Yes it is a bug in the current implementation. https://github.com/elastic/elasticsearch/pull/33518 should have added these bits as well.

Reply:
ok can we please split this into two PRs then? One for addressing the bug in the original implementation, and the new tests, the other one for adapting the security resolution code? That would make reviewing much easier.

Reply:
Okay, will do!